### PR TITLE
Implement MAP Policy Optimization (MPO)

### PR DESCRIPTION
**Pull request by TheodoreWolf**
**Stacked on https://github.com/keraJLi/rejax/pull/27, do not merge until previous PR is merged**

Implements [MPO](https://arxiv.org/abs/1806.06920) in rejax. The code was adapted from the [acme](https://github.com/google-deepmind/acme/blob/master/acme/jax/losses/mpo.py) version. MPO can be seen as a different version of SAC; they have the same policy/Q-critic with buffers, but different update algorithms.

Main differences from the original implementation:
- We use a scalar critic rather than a distributional one.
- We do a 1-step TD update rather than the default 5.
- We don't calculate any diagnostics.

I've also done some hyperparameter tuning for brax. The runs are public on [wandb](https://wandb.ai/flair/rejax-benchmark). I've also added an example, but granted it's not super useful, happy to remove.

### CODE REVIEW

File: examples/mpo_pendulum.py
Comment (keraJLi) [inline examples/mpo_pendulum.py:1]: I think this example isn't strictly needed. Maybe we can get away without it?

File: src/rejax/networks.py [hunk 1]
```diff
@@ -109,6 +109,41 @@ def action_log_prob(self, obs, rng):
         return action, action_dist.log_prob(action)
 
 
+class ClippedGaussianPolicy(nn.Module):
```
Comment (keraJLi) [inline src/rejax/networks.py]: While I understand that this class is very convenient for use with MPO specifically, I believe there is a little too much overlap with `GaussianPolicy`.
In PPO, the actions also need to be clipped, but this is done within the algorithm itself, and with good reason: clipping actions removes mass from the distribution function and makes logprobs ill-defined, which destabilizes training. This class also returns logprobs on the `__call__`, notably without a clipped action for which the logprob is accurate. But that actions are only clipped in `act` is a bit too confusing in my mind. 
Overall I think it might be best to use the `GaussianPolicy` for MPO, and apply action clipping within `MPO` itself, which should only be one extra line of code. Let me know what you think!

File: src/rejax/networks.py [hunk 2]
```diff
@@ -268,6 +303,21 @@ def __call__(self, obs, action):
         return nn.Dense(1)(x).squeeze(1)
 
 
+class ClippedQNetwork(nn.Module):
```
Comment (keraJLi) [inline src/rejax/networks.py]: With the changes suggested above (clipping actions in `MPO`) this class should not be needed

File: src/rejax/algos/mpo.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+                _MIN_LOG_TEMPERATURE, params["log_temperature"]
+            ),
+            "log_alpha_mean": jnp.maximum(_MIN_LOG_ALPHA, params["log_alpha_mean"]),
+            "log_alpha_stddev": jnp.maximum(_MIN_LOG_ALPHA, params["log_alpha_stddev"]),
+        }
+        if "log_penalty_temperature" in params:
+            clipped["log_penalty_temperature"] = jnp.maximum(
+                _MIN_LOG_TEMPERATURE, params["log_penalty_temperature"]
+            )
+        return FrozenDict(clipped)
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: I believe that in line with the rest of the code this should be a namedtuple

File: src/rejax/algos/mpo.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+            )
+            next_actions = target_dist.sample(
+                seed=action_rng, sample_shape=(self.policy_eval_num_val_samples,)
+            )
+
+            qs_next = jax.vmap(
+                lambda a: self.critic.apply(ts.critic_target_params, mb.next_obs, a)
+            )(next_actions)
+            q_target = jnp.mean(qs_next, axis=0)
+            target = jax.lax.stop_gradient(
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: This stop gradient is not required and should be removed for simplicity

File: src/rejax/algos/mpo.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+            optax.clip(self.max_grad_norm),
+            optax.adam(learning_rate=self.dual_learning_rate),
+        )
+        actor_ts = TrainState.create(apply_fn=(), params=actor_params, tx=tx)
+        critic_ts = TrainState.create(apply_fn=(), params=critic_params, tx=tx)
+
+        action_dim = int(np.prod(self.env.action_space(self.env_params).shape))
+        dual_shape = (action_dim,) if self.per_dim_constraining else (1,)
+        dual_params = {
+            "log_temperature": jnp.full((1,), self.init_log_temperature),
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: having shape `()` here would save ourselves a squeeze later.

Comment (keraJLi) [reply src/rejax/algos/mpo.py]: I can't see any point in the code where the singleton dimension is required. Maybe this could just be changed?

File: src/rejax/algos/mpo.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+            "log_alpha_mean": jnp.full(dual_shape, self.init_log_alpha_mean),
+            "log_alpha_stddev": jnp.full(dual_shape, self.init_log_alpha_stddev),
+        }
+        if self.action_penalization:
+            dual_params["log_penalty_temperature"] = jnp.full(
+                (1,), self.init_log_temperature
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: Here too could the singleton dimension be eliminated i believe

File: src/rejax/algos/mpo.py [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+
+        target_mean = target_dist.loc
+        target_scale = target_dist.scale_diag
+
+        def actor_and_dual_loss(actor_params, dual_params):
+            online_dist = self.actor.apply(actor_params, mb.obs, method="_action_dist")
+            online_mean = online_dist.loc  # [B, D]
+            online_scale = online_dist.scale_diag  # [B, D]
+
+            temperature = (
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: A lot of lines could be saved here by going `sp = jax.nn.softplus` and changing `_MPO_FLOAT_EPSILON` to `_MPO_EPS` or something

File: src/rejax/algos/mpo.py [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+    num_epochs: int = struct.field(pytree_node=False, default=1)
+    # N actions sampled per state in the E-step
+    num_action_samples: int = struct.field(pytree_node=False, default=20)
+    # N actions sampled per next state for critic value estimation
+    policy_eval_num_val_samples: int = struct.field(pytree_node=False, default=128)
+    # Enforce KL constraint per action dimension rather than jointly
+    per_dim_constraining: bool = struct.field(pytree_node=False, default=True)
+    # MO-MPO action penalization from the Acme loss.
+    action_penalization: bool = struct.field(pytree_node=False, default=True)
+    target_update_freq: int = struct.field(pytree_node=False, default=100)
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: Why are these here? These should be part of the TargetNetworkMixin

File: src/rejax/algos/mpo.py [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+        ts = ts.replace(
+            actor_ts=ts.actor_ts.apply_gradients(grads=actor_grad),
+            dual_ts=ts.dual_ts.apply_gradients(grads=dual_grad),
+        )
+        return ts.replace(
+            dual_ts=ts.dual_ts.replace(params=self.clip_dual_params(ts.dual_ts.params))
+        )
+
+    def update_target_networks(self, ts):
+        learner_step = ts.learner_step + 1
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: Any algorithm that has the `TargetNetworkMixin` should behave correctly when either `target_update_freq` or `polyak` is given. I think introducing the `learner_step` makes it pretty unclear if this is the case. On the first read it seems like the unit of `target_update_freq` is not `global_step / iteration_size`?

File: src/rejax/algos/mpo.py [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+            critic_target_params=self.update_target_params(
+                ts.critic_ts.params, ts.critic_target_params, learner_step
+            ),
+            learner_step=learner_step,
+        )
+
+    def update(self, ts, mb):
+        ts = self.update_critic(ts, mb)
+        ts = self.update_actor_and_duals(ts, mb)
+        # NOTE: bit of an overload, we are reusing the update_target_networks method
```
Comment (keraJLi) [inline src/rejax/algos/mpo.py]: I don't understand exactly what the comment is trying to say; what is overloaded?

**Review by keraJLi — CHANGES_REQUESTED** [review]
This PR includes changes from the previous PR on the target network mixin refactor, which should be disentangled.
